File: app/src/main/java/eu/darken/bluemusic/core/VolumeObserver.java
package eu.darken.bluemusic.core;

import android.database.ContentObserver;
import android.media.AudioManager;
import android.os.Handler;
import android.util.SparseArray;
import android.util.SparseIntArray;

import timber.log.Timber;


public class VolumeObserver extends ContentObserver {

    private final AudioManager audioManager;

    interface Callback {
        void onVolumeChanged(int streamId, int volume);
    }

    private final SparseArray<Callback> callbacks = new SparseArray<>();
    private final SparseIntArray volumes = new SparseIntArray();

    public VolumeObserver(Handler handler, AudioManager audioManager) {
        super(handler);
        this.audioManager = audioManager;
    }

    public void addCallback(int streamType, Callback callback) {
        callbacks.put(streamType, callback);
        final int volume = audioManager.getStreamVolume(streamType);
        volumes.put(streamType, volume);
    }

    @Override
    public boolean deliverSelfNotifications() {
        return super.deliverSelfNotifications();
    }

    @Override
    public void onChange(boolean selfChange) {
        super.onChange(selfChange);
        Timber.v("Change detected.");
        for (int i = 0; i < callbacks.size(); i++) {
            int streamType = callbacks.keyAt(i);
            Callback callback = callbacks.get(streamType);
            int newVolume = audioManager.getStreamVolume(streamType);
            int oldVolume = volumes.get(streamType);
            if (newVolume != oldVolume) {
                Timber.v("Volume changed (type=%d, old=%d, new=%d)", streamType, oldVolume, newVolume);
                volumes.put(streamType, newVolume);
                callback.onVolumeChanged(streamType, newVolume);
            }
        }
    }
}

File: app/src/main/java/eu/darken/bluemusic/monitor/core/audio/VolumeObserver.kt
package eu.darken.bluemusic.monitor.core.audio

import android.content.ContentResolver
import android.content.Context
import android.database.ContentObserver
import android.os.Handler
import android.os.HandlerThread
import android.provider.Settings
import dagger.hilt.android.qualifiers.ApplicationContext
import eu.darken.bluemusic.common.coroutine.AppScope
import eu.darken.bluemusic.common.debug.logging.Logging.Priority.VERBOSE
import eu.darken.bluemusic.common.debug.logging.log
import eu.darken.bluemusic.common.debug.logging.logTag
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.channels.trySendBlocking
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.shareIn
import java.util.concurrent.ConcurrentHashMap
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class VolumeObserver @Inject constructor(
    @ApplicationContext private val context: Context,
    @AppScope private val appScope: CoroutineScope,
    private val volumeTool: VolumeTool,
) {

    private val contentResolver: ContentResolver = context.contentResolver
    private val handler by lazy {
        val handlerThread = HandlerThread("VolumeObserver")
        handlerThread.start()
        val looper = handlerThread.looper
        Handler(looper)
    }

    private val volumesCache = ConcurrentHashMap<AudioStream.Id, Int>()

    val volumes: Flow<VolumeEvent> = callbackFlow {
        AudioStream.Id.entries.forEach { id ->
            val volume = volumeTool.getCurrentVolume(id)
            volumesCache[id] = volume
        }

        val observer = object : ContentObserver(handler) {

            override fun deliverSelfNotifications(): Boolean = true

            override fun onChange(selfChange: Boolean) {
                log(TAG, VERBOSE) { "Change detected (selfChange=$selfChange)" }
                AudioStream.Id.entries.forEach { id ->
                    val newVolume = volumeTool.getCurrentVolume(id)
                    val oldVolume = volumesCache[id] ?: -1
                    if (newVolume != oldVolume) {
                        val isSelf = volumeTool.wasUs(id, newVolume)
                        log(TAG) { "Volume changed (type=$id, old=$oldVolume, new=$newVolume, self=$isSelf)" }
                        volumesCache[id] = newVolume
                        val change = VolumeEvent(
                            streamId = id,
                            oldVolume = oldVolume,
                            newVolume = newVolume,
                            self = isSelf
                        )
                        trySendBlocking(change)
                    }
                }
            }
        }

        contentResolver.registerContentObserver(Settings.System.CONTENT_URI, true, observer)
        log(TAG) { "Now listening for volume change events" }

        awaitClose {
            log(TAG) { "Stopping listening for volume change events" }
            contentResolver.unregisterContentObserver(observer)
        }
    }.shareIn(
        scope = appScope,
        started = SharingStarted.WhileSubscribed(),
        replay = 0,
    )

    companion object {
        private val TAG = logTag("Monitor", "VolumeObserver")
    }
}